### PR TITLE
fix(session): prevent session conflict for API requests with portal c…

### LINKED ISSUE #10416: fix(session): API requests with portal cookies cause session conflict error
## Bug Description

When an API request comes in with a portal App cookie, `SessionWrapperFactory` tries to start a portal session even though an API session is already active (started by `SiteSetupListener`), causing a "Failed to start the session: already started by PHP" error.

## Steps to Reproduce

1. Make an API request that includes a portal app cookie
2. Session conflict occurs because:
   - `SiteSetupListener` starts an API session
   - `SessionWrapperFactory::findSessionWrapper()` detects the portal cookie
   - It attempts to start a portal session, but PHP session is already active

## Expected Behavior

API requests should use their existing Symfony sessions regardless of any portal cookies present.

## Actual Behavior

Error: "Failed to start the session: already started by PHP"

## Related Files

- `src/Common/Session/PHPSessionWrapper.php`
- `src/Common/Session/SessionWrapperFactory.php`
- `src/RestControllers/Subscriber/SiteSetupListener.php`
- `interface/globals.php`

### PR DESCRIPTION
**Pull request by rollingventures**
Fixes #10416

#### Short description of what this resolves:

When an API request comes in with a portal App cookie, `SessionWrapperFactory` tries to start a portal session even though an API session is already active (started by `SiteSetupListener`), causing a "Failed to start the session: already started by PHP" error.

### `src/Common/Session/PHPSessionWrapper.php`
- Added check in `__construct()` to detect API/OAuth requests before starting a core session
- If an API/OAuth request is detected, the constructor returns early without starting a session
- Prevents session conflicts when API sessions are already active

### `src/Common/Session/SessionWrapperFactory.php`
- Added check in `findSessionWrapper()` to detect API/OAuth requests before attempting to start a portal session
- If an API/OAuth request is detected, returns a `PHPSessionWrapper` instance without starting a new session
- Ensures API/OAuth requests use their existing Symfony sessions

## Technical Details

The fix uses `SessionUtil::API_WEBROOT` (`/apis/`) and `SessionUtil::OAUTH_WEBROOT` (`/oauth2/`) constants to detect API/OAuth requests by checking the `REQUEST_URI`. This prevents the session wrapper from attempting to start a new session when one is already active.

## Testing Considerations

- [x] API requests with portal cookies should no longer cause session conflicts
- [x] API requests without portal cookies should continue to work
- [x] Portal requests should continue to work normally
- [x] OAuth requests should continue to work normally
- [x] No session conflicts should occur in any scenario

## Related Files

- `src/Common/Session/SessionUtil.php` - Contains `API_WEBROOT` and `OAUTH_WEBROOT` constants
- `src/RestControllers/Subscriber/SiteSetupListener.php` - Starts API/OAuth sessions
- `interface/globals.php` - Calls `SessionWrapperFactory->getWrapper()`

### CODE REVIEW

**Comment by kojiromike** [comment]
@zmilan for review

**Comment by zmilan** [comment]
> @zmilan for review

This is good... I am sorry for not considering this when I was working on it. I didn't have experience with the API part or how to test it properly. But this is correct, it will start a regular wrapper around `$_SESSION`, and everything should work correctly.

**Comment by zmilan** [comment]
@rollingventures Not related to these changes, but to my work on moving the Core session to Symfony Session...what would be the way to properly initiate an API session in some of these shared files? I see, based on your changes, how to check if it is an API request...

File: src/Common/Session/PHPSessionWrapper.php [hunk 1]
```diff
@@ -22,6 +22,17 @@ class PHPSessionWrapper implements SessionWrapperInterface
 {
     public function __construct()
     {
+        // Check if we're in an API or OAuth request context
+        // If so, don't start a core session as API requests use Symfony sessions
+        $requestUri = $_SERVER['REQUEST_URI'] ?? '';
```
Comment (adunsulag) [inline src/Common/Session/PHPSessionWrapper.php]: Wouldn't it make more sense just to remove the webroot check on line 38 here and just return false if the session is already started? We'll need to default webroot to empty string due to type safety further down the call chain, but we don't want sessions getting started if there's an already existing session anyways.

It also simplifies the logic here.

If there's a compelling reason why we don't want to do this, then this code looks fine.

File: src/Common/Session/SessionWrapperFactory.php [hunk 2]
```diff
@@ -35,6 +35,21 @@ public function getWrapper(array $initData = []): SessionWrapperInterface
...
+        // Check if we're already in an API/OAuth session context
+        $requestUri = $_SERVER['REQUEST_URI'] ?? '';
+        $isApiRequest = strpos($requestUri, SessionUtil::API_WEBROOT) !== false || 
+                        strpos($requestUri, SessionUtil::OAUTH_WEBROOT) !== false;
+        
+        if ($isApiRequest) {
+            // API/OAuth requests already have Symfony sessions active
+            // Return a PHPSessionWrapper that won't start a new session
+            $session = new PHPSessionWrapper();
+            foreach ($initData as $name => $value) {
```
Comment (adunsulag) [inline src/Common/Session/SessionWrapperFactory.php]: This line is a duplicate of lines 48-50.  The if clause could be moved down to be part of the branching.

Line 39 could become:
if ($isApiRequest || $app !== SessionUtil::PORTAL_SESSION_ID) {

Comment (adunsulag) [reply src/Common/Session/SessionWrapperFactory.php]: I'm working on the isPredisSession logic stuff anyways so will need to bring in your changes before my fixes go in.

**Review by adunsulag — CHANGES_REQUESTED** [review]

File: src/Common/Session/PHPSessionWrapper.php (hunk 1 above)
Comment (rollingventures) [reply src/Common/Session/PHPSessionWrapper.php]: agreed...

- Removed the $webroot !== null check on line 38
- Added early return if a session is already active (prevents starting a new session)
- Defaulted $webroot to an empty string using the null coalescing operator (??) for type safety
- Simplified the logic: check API request → check if session active → start session with defaulted webroot

File: src/Common/Session/SessionWrapperFactory.php (hunk 2 above)
Comment (rollingventures) [reply src/Common/Session/SessionWrapperFactory.php]: removed, and doing actual testing with various use cases now.

- Consolidated duplicate code by combining the API request check with the non-portal check
- Changed line 45 to: if ($isApiRequest || $app !== SessionUtil::PORTAL_SESSION_ID)
- Moved the initData loop to after the branching logic, eliminating duplication between lines 47-49 and 63-65
- All three branches (API/non-portal, portal with Predis, portal with Symfony) now share the same initData setting logic

Comment (adunsulag) [reply src/Common/Session/SessionWrapperFactory.php]: I'll be dropping the predis check with my changes once you've finished this.  Predis logic is being moved into SessionUtil to be part of the storage engine setup.  Current implementation breaks with predis.  I'm tempted to remove the predis capability anyways and deal with it in our subsequent session migration that zmilan is doing, but I'm close to a resolution so probably going to stick with this. Thanks for taking this on.

**Comment by rollingventures** [comment]
@adunsulag @kojiromike - finishing many functional local session use case tests, php stan'sons and then this will be ready for final review.

**Comment by rollingventures** [comment]
> > @zmilan for review
> 
> This is good... I am sorry for not considering this when I was working on it. I didn't have experience with the API part or how to test it properly. But this is correct, it will start a regular wrapper around `$_SESSION`, and everything should work correctly.

No problem - the wrapper abstraction you built made this an easy fix. Thanks for confirming!

**Comment by rollingventures** [comment]
Included @adunsulag suggestions, and went with -> str_starts_with 
```
All 4 session types verified:
┌────────┬──────────────────────────────┬────────┐
│  Type  │          Cookie Set          │ Status │
├────────┼──────────────────────────────┼────────┤
│ API    │ apiOpenEMR only              │ ✅ OK  │
├────────┼──────────────────────────────┼────────┤
│ OAuth  │ authserverOpenEMR only       │ ✅ OK  │
├────────┼──────────────────────────────┼────────┤
│ Core   │ OpenEMR (+ App cookie)       │ ✅ OK  │
├────────┼──────────────────────────────┼────────┤
│ Portal │ PortalOpenEMR (+ App cookie) │ ✅ OK  │
└────────┴──────────────────────────────┴────────┘
 ```
 With some extra analysis:
 ```
  ┌──────────────┬───────────────────┬──────────┬─────────────────────────────────────────────────────────────────────┐
  │ Session Type │    Cookie Name    │   Path   │                          How It's Handled                           │
  ├──────────────┼───────────────────┼──────────┼─────────────────────────────────────────────────────────────────────┤
  │ Core UI      │ OpenEMR           │ /        │ PHPSessionWrapper starts session only if none active                │
  ├──────────────┼───────────────────┼──────────┼─────────────────────────────────────────────────────────────────────┤
  │ API          │ apiOpenEMR        │ /apis/   │ Detected by path → PHPSessionWrapper skips (session already active) │
  ├──────────────┼───────────────────┼──────────┼─────────────────────────────────────────────────────────────────────┤
  │ OAuth        │ authserverOpenEMR │ /oauth2/ │ Detected by path → PHPSessionWrapper skips (session already active) │
  ├──────────────┼───────────────────┼──────────┼─────────────────────────────────────────────────────────────────────┤
  │ Portal       │ PortalOpenEMR     │ /portal/ │ Detected by App cookie → SymfonySessionWrapper                      │
  ├──────────────┼───────────────────┼──────────┼─────────────────────────────────────────────────────────────────────┤
  │ Portal+Redis │ PortalOpenEMR     │ /portal/ │ Uses PHPSessionWrapper after Predis init                            │
  └──────────────┴───────────────────┴──────────┴─────────────────────────────────────────────────────────────────────┘
```

**Review by adunsulag — APPROVED** [review]
This looks fine to me with the changes.  If @kojiromike wants to take another look, but should be able to bring this in tomorrow morning.

File: src/Common/Session/SessionWrapperFactory.php [hunk 3]
```diff
@@ -35,11 +35,20 @@ public function getWrapper(array $initData = []): SessionWrapperInterface
 
     private function findSessionWrapper(array $initData = []): SessionWrapperInterface
     {
+        // Check if we're already in an API/OAuth session context
+        $requestUri = $_SERVER['REQUEST_URI'] ?? '';
+        $isApiRequest = str_starts_with((string) $requestUri, SessionUtil::API_WEBROOT) ||
```
Comment (kojiromike) [inline src/Common/Session/SessionWrapperFactory.php]: Will this still work correctly if openemr is installed at a subpath, like `http://example.com/openemr/apis/foo`

File: src/Common/Session/PHPSessionWrapper.php [hunk 4]
```diff
@@ -22,11 +22,14 @@ class PHPSessionWrapper implements SessionWrapperInterface
...
-        if ($webroot !== null && session_status() !== PHP_SESSION_ACTIVE) {
-            SessionUtil::coreSessionStart($webroot, false);
+        // Don't start a session if one is already active
+        if (session_status() === PHP_SESSION_ACTIVE) {
+            return;
         }
+
+        $globalsBag = OEGlobalsBag::getInstance();
+        $webroot = $globalsBag->get('webroot') ?? '';
+        SessionUtil::coreSessionStart($webroot, false);
```
Comment (kojiromike) [inline src/Common/Session/PHPSessionWrapper.php:33]: Does coreSessionStart do the right thing if `$webroot === ''`/ if the global webroot was not properly initialized, and is null?

File: src/Common/Session/SessionWrapperFactory.php [hunk 5]
```diff
@@ -35,11 +35,20 @@ public function getWrapper(array $initData = []): SessionWrapperInterface
...
+                        str_starts_with((string) $requestUri, SessionUtil::OAUTH_WEBROOT);
+
         $app = SessionUtil::getAppCookie();
-        if (SessionUtil::isPredisSession()) {
-            SessionUtil::portalPredisSessionStart();
+
+        if ($isApiRequest || $app !== SessionUtil::PORTAL_SESSION_ID) {
```
Comment (kojiromike) [inline src/Common/Session/SessionWrapperFactory.php]: This would be broken if API requests used predis. IIUC you did this because they don't, but please comment the code so it's clear why this order of precedence is ok.

File: src/Common/Session/SessionWrapperFactory.php [hunk 6]
```diff
@@ -35,11 +35,20 @@ public function getWrapper(array $initData = []): SessionWrapperInterface
...
+            // API/OAuth requests already have Symfony sessions active
+            // Non-portal requests use PHPSessionWrapper
+            // Return a PHPSessionWrapper that won't start a new session
             $session = new PHPSessionWrapper();
-        } else if ($app !== SessionUtil::PORTAL_SESSION_ID) {
```
Comment (kojiromike) [inline src/Common/Session/SessionWrapperFactory.php:42]: 🤔 could the entire change actually just boil down to adding `|| session_status() === PHP_SESSION_ACTIVE` to this `else if`?

Comment (rollingventures) [reply src/Common/Session/SessionWrapperFactory.php:42]: Yes - much cleaner than path detection.  Modified, thank you.

File: src/Common/Session/SessionWrapperFactory.php (hunk 5 above)
Comment (rollingventures) [reply src/Common/Session/SessionWrapperFactory.php]: This is now moot - I removed the path detection entirely and went with your session_status() suggestion instead. The predis check remains first as before.

File: src/Common/Session/PHPSessionWrapper.php (hunk 4 above)
Comment (rollingventures) [reply src/Common/Session/PHPSessionWrapper.php:33]: Yes - empty webroot is valid and means OpenEMR is installed at the document root. The cookie path defaults to / in that case. Added a comment clarifying this.

File: src/Common/Session/SessionWrapperFactory.php (hunk 3 above)
Comment (rollingventures) [reply src/Common/Session/SessionWrapperFactory.php]: Removed the path detection entirely. The fix now relies solely on session_status() === PHP_SESSION_ACTIVE, which is path-agnostic. Added tests for root, subpath (/openemr/apis/), and deep subpath (/health/systems/openemr/apis/) installations - all pass.

**Comment by rollingventures** [comment]
Also added tests, and will conduct some more custom edge cases before this is ready to merge pending ci agrees also.
```
Tests added (SessionWrapperFactoryTest.php - 20 tests):
┌─────────────────┬────────────────────────────────────────────────────────┐
│    Category     │                        Coverage                        │
├─────────────────┼────────────────────────────────────────────────────────┤
│ Session Types   │ Core, API, OAuth, Portal                               │
├─────────────────┼────────────────────────────────────────────────────────┤
│ Shared Files    │ Portal cookie + API/OAuth session, module entry points │
├─────────────────┼────────────────────────────────────────────────────────┤
│ Path Variations │ Root, subpath, deep subpath, empty/missing URI         │
├─────────────────┼────────────────────────────────────────────────────────┤
│ Edge Cases      │ Missing cookie, cached wrapper, init data              │
└─────────────────┴────────────────────────────────────────────────────────┘
OK (20 tests, 21 assertions)
Full suite: 425 tests ✅
```

**Comment by rollingventures** [comment]
@adunsulag -> Predis edge case - If isPredisSession() is true AND a session is already active,portalPredisSessionStart() gets called before our check. Could conflict in Redis environments with the same portal-cookie-during-API scenario. Likely rare but worth noting.  I know you did mention you were working on this logic so this PR is not blocked.

Also a potential "Portal" behavior change that I am researching today - If session is already active with portal cookie, we now return PHPSessionWrapper instead of SymfonySessionWrapper.

Testing approach:

Test: 1. Portal smoke test

How: Log into patient portal, test messaging, signatures, payments - verify isSymfonySession() -> returns true

Test: 2. API + portal cookie

How: Open portal in tab 1, make API call in tab 2, verify no session conflict

Test: 3. Check entry points

How: Verify no portal files start sessions before calling SessionWrapperFactory

**Comment by rollingventures** [comment]
Manual test confirmed:

- ✅ Portal login works normally
- ✅ API + portal cookie: returns 401 (expected), no session conflict
- ✅ Without fix: 500 - "Failed to start the session: already started by PHP."
- ✅ With fix: Clean 401, no session errors

29 unit tests covering all session types, shared file patterns, and path variations.

**Comment by adunsulag** [comment]
@rollingventures thanks, there's a number of problems in the predis stuff I'm working on.  I may just disable the predis altogether if I don't have a solution in the next hour or two.  I appreciate you getting this in.
